File: ts/src/data/operator.ts
/**
 * Operator data reader — loads and formats game data from local JSON files.
 * Mirrors python/src/prts_mcp/data/operator.py.
 *
 * JSON files are large (character_table.json ~4 MB) so they are loaded
 * lazily on first call and cached in module-level variables.
 */

import { loadConfig, hasOperatorData } from "../config.js";
import { DirectoryStore } from "./stores.js";
import { stripWikitext } from "../utils/sanitizer.js";

// ---------------------------------------------------------------------------
// Module-level lazy caches
// ---------------------------------------------------------------------------

// Config is NOT cached here: loadConfig() re-checks file existence on each
// call, so effectiveExcelPath correctly reflects data written by auto-sync
// after startup. The cost is negligible (env-var reads + existsSync calls).

// Raw table caches — null means "not yet loaded", undefined means "failed".
type TableCache<T> = T | null | undefined;

let _characterTable: TableCache<Record<string, CharacterEntry>> = null;
let _handbookTable: TableCache<HandbookTable> = null;
let _charwordTable: TableCache<CharwordTable> = null;
let _nameToId: Map<string, string> | null = null;

export function clearOperatorCaches(): void {
  _characterTable = null;
  _handbookTable = null;
  _charwordTable = null;
  _nameToId = null;
}

// ---------------------------------------------------------------------------
// JSON shape types (only the fields we actually use)
// ---------------------------------------------------------------------------

interface CharacterEntry {
  name?: string;
  appellation?: string;
  displayNumber?: string;
  description?: string;
  rarity?: string;
  profession?: string;
  subProfessionId?: string;
  position?: string;
  nationId?: string;
  groupId?: string;
  teamId?: string;
  tagList?: string[];
  itemUsage?: string;
  itemDesc?: string;
  itemObtainApproach?: string;
  talents?: TalentSlot[];
}

interface TalentCandidate {
  name?: string;
  description?: string;
}

interface TalentSlot {
  candidates?: TalentCandidate[];
}

interface StoryEntry {
  storyTitle?: string;
  stories?: Array<{ storyText?: string }>;
}

interface HandbookEntry {
  storyTextAudio?: StoryEntry[];
}

interface HandbookTable {
  handbookDict?: Record<string, HandbookEntry>;
}

interface CharwordEntry {
  charId?: string;
  voiceTitle?: string;
  voiceText?: string;
}

interface CharwordTable {
  charWords?: Record<string, CharwordEntry>;
}

// ---------------------------------------------------------------------------
// Internal helpers
// ---------------------------------------------------------------------------

function missingDataMessage(): string {
  const cfg = loadConfig();
  return (
    "干员数据暂不可用。" +
    "容器启动时的 auto-sync 可能仍在进行中，请稍后重试；" +
    "若持续出现此提示，请检查网络连接或提供 GITHUB_TOKEN 以降低限速风险。" +
    `（当前同步目标路径：${cfg.excelPath}）`
  );
}

function loadJson<T>(filePath: string): T {
  const store = operatorStore();
  if (!store.exists(filePath)) {
    throw new Error(
      `干员数据文件不存在：${store.resolveForDiagnostics(filePath)}。` +
        "数据目录可能为空，或挂载路径有误（GAMEDATA_PATH 应指向 ArknightsGameData 仓库根目录）。"
    );
  }
  return store.readJson<T>(filePath);
}

function operatorStore(): DirectoryStore {
  const ep = loadConfig().effectiveExcelPath;
  if (ep === null) throw new Error("effectiveExcelPath is null");
  return new DirectoryStore(ep);
}

export function getCharacterTable(): Record<string, CharacterEntry> {
  if (_characterTable === null) {
    _characterTable = loadJson<Record<string, CharacterEntry>>(
      "character_table.json"
    );
  }
  if (_characterTable === undefined) throw new Error("character_table failed");
  return _characterTable;
}

export function getHandbookTable(): HandbookTable {
  if (_handbookTable === null) {
    _handbookTable = loadJson<HandbookTable>(
      "handbook_info_table.json"
    );
  }
  if (_handbookTable === undefined) throw new Error("handbook_table failed");
  return _handbookTable;
}

export function getCharwordTable(): CharwordTable {
  if (_charwordTable === null) {
    _charwordTable = loadJson<CharwordTable>("charword_table.json");
  }
  if (_charwordTable === undefined) throw new Error("charword_table failed");
  return _charwordTable;
}

export function resolveCharId(name: string): string | null {
  if (_nameToId === null) {
    const ct = getCharacterTable();
    _nameToId = new Map(
      Object.entries(ct)
        .filter(([cid, info]) => info.name && cid.startsWith("char_"))
        .map(([cid, info]) => [info.name!, cid])
    );
  }
  return _nameToId.get(name) ?? null;
}

// ---------------------------------------------------------------------------
// Public API
// ---------------------------------------------------------------------------

/** Return formatted archive text for an operator by Chinese name. */
export function getOperatorArchives(name: string): string {
  const cfg = loadConfig();
  if (!hasOperatorData(cfg)) return missingDataMessage();

  let charId: string | null;
  try {
    charId = resolveCharId(name);
  } catch (err) {
    return err instanceof Error ? err.message : String(err);
  }
  if (charId === null) {
    return `未找到干员 '${name}'。请使用游戏内中文名称（如'阿米娅'）。`;
  }

  let handbook: HandbookTable;
  try {
    handbook = getHandbookTable();
  } catch (err) {
    return err instanceof Error ? err.message : String(err);
  }

  const entry = handbook.handbookDict?.[charId];
  if (!entry) return `干员 '${name}' 暂无档案数据。`;

  const sections: string[] = [];
  for (const story of entry.storyTextAudio ?? []) {
    const title = story.storyTitle ?? "";
    const texts = (story.stories ?? [])
      .map((s) => s.storyText ?? "")
      .filter(Boolean);
    if (texts.length > 0) {
      sections.push(`### ${title}\n` + texts.join("\n"));
    }
  }

  if (sections.length === 0) return `干员 '${name}' 档案内容为空。`;
  return `# ${name} - 干员档案\n\n` + sections.join("\n\n");
}

/** Return formatted voice-line text for an operator by Chinese name. */
export function getOperatorVoicelines(name: string): string {
  const cfg = loadConfig();
  if (!hasOperatorData(cfg)) return missingDataMessage();

  let charId: string | null;
  try {
    charId = resolveCharId(name);
  } catch (err) {
    return err instanceof Error ? err.message : String(err);
  }
  if (charId === null) {
    return `未找到干员 '${name}'。请使用游戏内中文名称（如'阿米娅'）。`;
  }

  let charwords: CharwordTable;
  try {
    charwords = getCharwordTable();
  } catch (err) {
    return err instanceof Error ? err.message : String(err);
  }

  const lines: string[] = [];
  for (const entry of Object.values(charwords.charWords ?? {})) {
    if (entry.charId === charId && entry.voiceText) {
      const title = entry.voiceTitle ?? "未知";
      lines.push(`**${title}**: ${entry.voiceText}`);
    }
  }

  if (lines.length === 0) return `干员 '${name}' 暂无语音数据。`;
  return `# ${name} - 语音记录\n\n` + lines.join("\n");
}

// ---------------------------------------------------------------------------
// Basic info
// ---------------------------------------------------------------------------

const PROFESSION_ZH: Record<string, string> = {
  CASTER: "术师",
  MEDIC: "医疗",
  PIONEER: "先锋",
  SNIPER: "狙击",
  SPECIAL: "特种",
  SUPPORT: "辅助",
  TANK: "重装",
  WARRIOR: "近卫",
};

const POSITION_ZH: Record<string, string> = {
  RANGED: "远程",
  MELEE: "近战",
  ALL: "通用",
  NONE: "-",
};

/** Return basic profile info for an operator by Chinese name. */
export function getOperatorBasicInfo(name: string): string {
  const cfg = loadConfig();
  if (!hasOperatorData(cfg)) return missingDataMessage();

  let charId: string | null;
  try {
    charId = resolveCharId(name);
  } catch (err) {
    return err instanceof Error ? err.message : String(err);
  }
  if (charId === null) {
    return `未找到干员 '${name}'。请使用游戏内中文名称（如'阿米娅'）。`;
  }

  let ct: Record<string, CharacterEntry>;
  try {
    ct = getCharacterTable();
  } catch (err) {
    return err instanceof Error ? err.message : String(err);
  }
  const info = ct[charId];
  if (!info) return `干员 '${name}' 暂无基本信息。`;

  const rarityRaw = info.rarity ?? "";
  const rarity = rarityRaw.startsWith("TIER_")
    ? rarityRaw.replace("TIER_", "") + "★"
    : rarityRaw;

  const profession = PROFESSION_ZH[info.profession ?? ""] ?? (info.profession ?? "");
  const position = POSITION_ZH[info.position ?? ""] ?? (info.position ?? "");

  const affiliationParts = [info.nationId, info.groupId, info.teamId].filter(Boolean) as string[];
  const affiliation = affiliationParts.length > 0 ? affiliationParts.join(" / ") : "-";

  const lines: string[] = [`# ${name} - 干员基本信息\n`];
  lines.push(`- **编号**：${info.displayNumber ?? ""}`);
  lines.push(`- **英文名**：${info.appellation ?? ""}`);
  lines.push(`- **稀有度**：${rarity}`);
  lines.push(`- **职业**：${profession}（${info.subProfessionId ?? ""}）`);
  lines.push(`- **站位**：${position}`);
  lines.push(`- **所属**：${affiliation}`);
  if (info.tagList && info.tagList.length > 0) {
    lines.push(`- **招募标签**：${info.tagList.join("、")}`);
  }
  if (info.description) {
    lines.push(`- **攻击属性**：${stripWikitext(info.description)}`);
  }
  if (info.itemUsage) {
    lines.push(`\n**图鉴**：${info.itemUsage}`);
  }
  if (info.itemDesc) {
    lines.push(`\n> ${info.itemDesc}`);
  }
  if (info.itemObtainApproach) {
    lines.push(`\n**获取方式**：${info.itemObtainApproach}`);
  }

  const talents = info.talents ?? [];
  if (talents.length > 0) {
    lines.push("\n## 天赋");
    for (const slot of talents) {
      const candidates = slot.candidates ?? [];
      let chosen: TalentCandidate | undefined;
      for (let i = candidates.length - 1; i >= 0; i--) {
        const c = candidates[i];
        if (c.name && c.name !== "？？？") {
          chosen = c;
          break;
        }
      }
      if (chosen) {
        lines.push(`- **${chosen.name ?? ""}**：${stripWikitext(chosen.description ?? "")}`);
      }
    }
  }

  return lines.join("\n");
}

File: python/src/prts_mcp/data/operator.py
from __future__ import annotations

from functools import lru_cache
from typing import Any

from prts_mcp.config import Config
from prts_mcp.data.stores import DirectoryStore
from prts_mcp.utils.sanitizer import strip_wikitext

def _get_config() -> Config:
    return Config.load()


def clear_operator_caches() -> None:
    """Clear lazy table caches after synced game data changes on disk."""
    _load_character_table.cache_clear()
    _load_handbook_table.cache_clear()
    _load_charword_table.cache_clear()
    _build_name_to_id.cache_clear()
    try:
        from prts_mcp.data.search import clear_search_caches

        clear_search_caches()
    except ImportError:
        pass


def _missing_operator_data_message() -> str:
    config = _get_config()
    searched = str(config.excel_path)
    return (
        "干员数据暂不可用。"
        "容器启动时的 auto-sync 可能仍在进行中，请稍后重试；"
        "若持续出现此提示，请检查网络连接或提供 GITHUB_TOKEN 以降低限速风险。"
        f"（当前同步目标路径：{searched}）"
    )


def _operator_store() -> DirectoryStore:
    ep = _get_config().effective_excel_path
    assert ep is not None
    return DirectoryStore(ep)


def _load_json(filename: str) -> dict[str, Any]:
    store = _operator_store()
    if not store.exists(filename):
        raise FileNotFoundError(
            f"干员数据文件不存在：{store.root / filename}。"
            "数据目录可能为空，或挂载路径有误（GAMEDATA_PATH 应指向 ArknightsGameData 仓库根目录）。"
        )
    return store.read_json(filename)


@lru_cache(maxsize=1)
def _load_character_table() -> dict[str, Any]:
    return _load_json("character_table.json")


@lru_cache(maxsize=1)
def _load_handbook_table() -> dict[str, Any]:
    return _load_json("handbook_info_table.json")


@lru_cache(maxsize=1)
def _load_charword_table() -> dict[str, Any]:
    return _load_json("charword_table.json")


@lru_cache(maxsize=1)
def _build_name_to_id() -> dict[str, str]:
    """Map operator Chinese name -> charId."""
    ct = _load_character_table()
    return {info["name"]: cid for cid, info in ct.items()
            if info.get("name") and cid.startswith("char_")}


def _resolve_char_id(name: str) -> str | None:
    mapping = _build_name_to_id()
    return mapping.get(name)


# ---------------------------------------------------------------------------
# Public API
# ---------------------------------------------------------------------------


def get_operator_archives(name: str) -> str:
    """Return formatted archive text for an operator by Chinese name."""
    if not _get_config().has_operator_data:
        return _missing_operator_data_message()

    try:
        char_id = _resolve_char_id(name)
    except FileNotFoundError as exc:
        return str(exc)
    if char_id is None:
        return f"未找到干员 '{name}'。请使用游戏内中文名称（如'阿米娅'）。"

    try:
        handbook = _load_handbook_table().get("handbookDict", {})
    except FileNotFoundError as exc:
        return str(exc)
    entry = handbook.get(char_id)
    if entry is None:
        return f"干员 '{name}' 暂无档案数据。"

    sections: list[str] = []
    for story in entry.get("storyTextAudio", []):
        title = story.get("storyTitle", "")
        texts = [s.get("storyText", "") for s in story.get("stories", []) if s.get("storyText")]
        if texts:
            sections.append(f"### {title}\n" + "\n".join(texts))

    if not sections:
        return f"干员 '{name}' 档案内容为空。"
    return f"# {name} - 干员档案\n\n" + "\n\n".join(sections)


def get_operator_voicelines(name: str) -> str:
    """Return formatted voice-line text for an operator by Chinese name."""
    if not _get_config().has_operator_data:
        return _missing_operator_data_message()

    try:
        char_id = _resolve_char_id(name)
    except FileNotFoundError as exc:
        return str(exc)
    if char_id is None:
        return f"未找到干员 '{name}'。请使用游戏内中文名称（如'阿米娅'）。"

    try:
        charwords = _load_charword_table().get("charWords", {})
    except FileNotFoundError as exc:
        return str(exc)
    lines: list[str] = []
    for entry in charwords.values():
        if entry.get("charId") == char_id and entry.get("voiceText"):
            title = entry.get("voiceTitle", "未知")
            lines.append(f"**{title}**: {entry['voiceText']}")

    if not lines:
        return f"干员 '{name}' 暂无语音数据。"
    return f"# {name} - 语音记录\n\n" + "\n".join(lines)


# ---------------------------------------------------------------------------
# Basic info helpers
# ---------------------------------------------------------------------------

_PROFESSION_ZH: dict[str, str] = {
    "CASTER": "术师",
    "MEDIC": "医疗",
    "PIONEER": "先锋",
    "SNIPER": "狙击",
    "SPECIAL": "特种",
    "SUPPORT": "辅助",
    "TANK": "重装",
    "WARRIOR": "近卫",
}

_POSITION_ZH: dict[str, str] = {
    "RANGED": "远程",
    "MELEE": "近战",
    "ALL": "通用",
    "NONE": "-",
}


def get_operator_basic_info(name: str) -> str:
    """Return basic profile info for an operator by Chinese name."""
    if not _get_config().has_operator_data:
        return _missing_operator_data_message()

    try:
        char_id = _resolve_char_id(name)
    except FileNotFoundError as exc:
        return str(exc)
    if char_id is None:
        return f"未找到干员 '{name}'。请使用游戏内中文名称（如'阿米娅'）。"

    try:
        ct = _load_character_table()
    except FileNotFoundError as exc:
        return str(exc)
    info = ct.get(char_id)
    if info is None:
        return f"干员 '{name}' 暂无基本信息。"

    rarity_raw: str = info.get("rarity", "")
    rarity = rarity_raw.replace("TIER_", "") + "★" if rarity_raw.startswith("TIER_") else rarity_raw

    profession_raw: str = info.get("profession", "")
    profession = _PROFESSION_ZH.get(profession_raw, profession_raw)

    position_raw: str = info.get("position", "")
    position = _POSITION_ZH.get(position_raw, position_raw)

    appellation: str = info.get("appellation", "")
    display_number: str = info.get("displayNumber", "")
    description: str = info.get("description", "")
    item_usage: str = info.get("itemUsage", "")
    item_desc: str = info.get("itemDesc", "")
    item_obtain: str = info.get("itemObtainApproach", "")
    tag_list: list[str] = info.get("tagList") or []

    nation_id: str = info.get("nationId") or ""
    group_id: str = info.get("groupId") or ""
    team_id: str = info.get("teamId") or ""
    affiliation_parts = [x for x in [nation_id, group_id, team_id] if x]
    affiliation = " / ".join(affiliation_parts) if affiliation_parts else "-"

    lines: list[str] = [f"# {name} - 干员基本信息\n"]
    lines.append(f"- **编号**：{display_number}")
    lines.append(f"- **英文名**：{appellation}")
    lines.append(f"- **稀有度**：{rarity}")
    lines.append(f"- **职业**：{profession}（{info.get('subProfessionId', '')}）")
    lines.append(f"- **站位**：{position}")
    lines.append(f"- **所属**：{affiliation}")
    if tag_list:
        lines.append(f"- **招募标签**：{'、'.join(tag_list)}")
    if description:
        lines.append(f"- **攻击属性**：{strip_wikitext(description)}")
    if item_usage:
        lines.append(f"\n**图鉴**：{item_usage}")
    if item_desc:
        lines.append(f"\n> {item_desc}")
    if item_obtain:
        lines.append(f"\n**获取方式**：{item_obtain}")

    # Talents — show the highest-unlock candidate for each talent slot
    talents: list[Any] = info.get("talents") or []
    if talents:
        lines.append("\n## 天赋")
        for talent in talents:
            candidates: list[Any] = talent.get("candidates") or []
            # Pick last candidate with a real name
            chosen = None
            for c in reversed(candidates):
                if c.get("name") and c["name"] not in ("？？？",):
                    chosen = c
                    break
            if chosen:
                t_name: str = chosen.get("name", "")
                t_desc: str = strip_wikitext(chosen.get("description", ""))
                lines.append(f"- **{t_name}**：{t_desc}")

    return "\n".join(lines)
